File: app/src/main/java/catchla/yep/activity/SignUpActivity.java
/*
 * Copyright 2012 Roman Nurik
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package catchla.yep.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.bluelinelabs.logansquare.LoganSquare;
import com.desmond.asyncmanager.AsyncManager;
import com.desmond.asyncmanager.TaskRunnable;

import java.io.IOException;

import catchla.yep.Constants;
import catchla.yep.R;
import catchla.yep.adapter.TabsAdapter;
import catchla.yep.fragment.ProgressDialogFragment;
import catchla.yep.model.AccessToken;
import catchla.yep.model.Client;
import catchla.yep.model.CreateRegistrationResult;
import catchla.yep.model.TaskResponse;
import catchla.yep.util.ParseUtils;
import catchla.yep.util.YepAPI;
import catchla.yep.util.YepAPIFactory;
import catchla.yep.util.YepException;

public class SignUpActivity extends ContentActivity implements Constants, ViewPager.OnPageChangeListener,
        View.OnClickListener {

    private ViewPager mViewPager;
    private TabsAdapter mAdapter;

    private Button mNextButton;
    private String mName;
    private CreateRegistrationResult mCreateRegistrationResult;
    private AccessToken mAccessToken;


    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_up);
        mAdapter = new TabsAdapter(this, getSupportFragmentManager());
        mViewPager.setAdapter(mAdapter);
        mViewPager.setEnabled(false);
        mViewPager.setOnPageChangeListener(this);

        mNextButton.setOnClickListener(this);

        mAdapter.addTab(EditNameFragment.class, null, 0, null);
        mAdapter.addTab(EditPhoneNumberFragment.class, null, 0, null);
        mAdapter.addTab(VerifyPhoneNumberFragment.class, null, 0, null);
        mAdapter.addTab(EditAvatarFragment.class, null, 0, null);
    }

    @Override
    public void onContentChanged() {
        super.onContentChanged();
        mViewPager = (ViewPager) findViewById(R.id.view_pager);
        mNextButton = (Button) findViewById(R.id.next_button);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

    }

    @Override
    public void onPageSelected(int position) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.next_button: {
                final Fragment fragment = mAdapter.getPrimaryItem();
                if (fragment instanceof AbsSignUpPageFragment) {
                    ((AbsSignUpPageFragment) fragment).onNextPage();
                }
                break;
            }
        }
    }

    private void setName(final String name) {
        mName = name;
    }

    private void setNextEnabled(boolean enabled) {
        mNextButton.setEnabled(enabled);
    }

    private void gotoNextPage() {
        final int currentItem = mViewPager.getCurrentItem();
        if (currentItem == mAdapter.getCount() - 1) return;
        mViewPager.setCurrentItem(currentItem + 1);
    }

    private void sendVerifyCode(final String phoneNumber, final String countryCode) {
        ProgressDialogFragment.show(this, "create_registration");
        final TaskRunnable<String[], TaskResponse<CreateRegistrationResult>, SignUpActivity> task
                = new TaskRunnable<String[], TaskResponse<CreateRegistrationResult>, SignUpActivity>() {

            @Override
            public TaskResponse<CreateRegistrationResult> doLongOperation(final String[] args) throws InterruptedException {
                final YepAPI yep = YepAPIFactory.getInstanceWithToken(SignUpActivity.this, null);
                try {
                    return TaskResponse.getInstance(yep.createRegistration(args[1], args[2], args[0], 0, 0));
                } catch (YepException e) {
                    return TaskResponse.getInstance(e);
                }
            }

            @Override
            public void callback(final SignUpActivity handler, final TaskResponse<CreateRegistrationResult> result) {
                final Fragment f = handler.getSupportFragmentManager().findFragmentByTag("create_registration");
                if (f instanceof DialogFragment) {
                    ((DialogFragment) f).dismiss();
                }
                if (result.hasData()) {
                    handler.setCreateRegistrationResult(result.getData());
                    handler.gotoNextPage();
                } else {
                    @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
                    final YepException exception = (YepException) result.getException();
                    final String error = exception.getError();
                    if (TextUtils.isEmpty(error)) {
                        Toast.makeText(handler, R.string.unable_to_register, Toast.LENGTH_SHORT).show();
                    } else {
                        Toast.makeText(handler, error, Toast.LENGTH_SHORT).show();
                    }
                }
            }

        };
        task.setParams(new String[]{mName, phoneNumber, countryCode});
        task.setResultHandler(this);
        AsyncManager.runBackgroundTask(task);
    }

    private void verifyPhoneNumber(final String verifyCode) {
        ProgressDialogFragment.show(this, "update_registration");
        final TaskRunnable<Pair<CreateRegistrationResult, String>, TaskResponse<AccessToken>, SignUpActivity> task
                = new TaskRunnable<Pair<CreateRegistrationResult, String>, TaskResponse<AccessToken>, SignUpActivity>() {

            @Override
            public TaskResponse<AccessToken> doLongOperation(final Pair<CreateRegistrationResult, String> args) throws InterruptedException {
                final YepAPI yep = YepAPIFactory.getInstanceWithToken(SignUpActivity.this, null);
                try {
                    final CreateRegistrationResult result = args.first;
                    final String verifyCode = args.second;
                    return TaskResponse.getInstance(yep.updateRegistration(result.getMobile(),
                            result.getPhoneCode(), verifyCode, Client.OFFICIAL, 0));
                } catch (YepException e) {
                    return TaskResponse.getInstance(e);
                }
            }

            @Override
            public void callback(final SignUpActivity handler, final TaskResponse<AccessToken> result) {
                final Fragment f = handler.getSupportFragmentManager().findFragmentByTag("update_registration");
                if (f instanceof DialogFragment) {
                    ((DialogFragment) f).dismiss();
                }
                if (result.hasData()) {
                    handler.setUpdateRegistrationResult(result.getData());
                    handler.gotoNextPage();
                } else {
                    @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
                    final YepException exception = (YepException) result.getException();
                    final String error = exception.getError();
                    if (TextUtils.isEmpty(error)) {
                        Toast.makeText(handler, R.string.unable_to_verify_phone, Toast.LENGTH_SHORT).show();
                    } else {
                        Toast.makeText(handler, error, Toast.LENGTH_SHORT).show();
                    }
                }
            }

        };
        task.setParams(Pair.create(mCreateRegistrationResult, verifyCode));
        task.setResultHandler(this);
        AsyncManager.runBackgroundTask(task);
    }

    private void setUpdateRegistrationResult(final AccessToken result) {
        mAccessToken = result;
    }

    public void setCreateRegistrationResult(final CreateRegistrationResult result) {
        mCreateRegistrationResult = result;
    }

    private abstract static class AbsSignUpPageFragment extends Fragment {
        protected abstract void updateNextButton();

        public abstract void onNextPage();

        @Override
        public void setUserVisibleHint(boolean isVisibleToUser) {
            super.setUserVisibleHint(isVisibleToUser);
            if (isVisibleToUser) {
                updateNextButton();
            }
        }

        public SignUpActivity getSignUpActivity() {
            return (SignUpActivity) getActivity();
        }

        @Override
        public void onActivityCreated(@Nullable Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);
            updateNextButton();
        }
    }

    public static class EditNameFragment extends AbsSignUpPageFragment {
        private EditText mEditName;

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.fragment_sign_up_edit_name, container, false);
        }

        @Override
        public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            mEditName = (EditText) view.findViewById(R.id.edit_name);
        }

        @Override
        public void onActivityCreated(@Nullable Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);
            mEditName.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    updateNextButton();
                }

                @Override
                public void afterTextChanged(Editable s) {

                }
            });
        }

        @Override
        protected void updateNextButton() {
            if (!getUserVisibleHint()) return;
            final SignUpActivity signUpActivity = getSignUpActivity();
            if (signUpActivity == null || mEditName == null) return;
            signUpActivity.setNextEnabled(mEditName.length() > 0);
        }

        @Override
        public void onNextPage() {
            final SignUpActivity signUpActivity = getSignUpActivity();
            signUpActivity.setName(ParseUtils.parseString(mEditName.getText()));
            signUpActivity.gotoNextPage();
        }
    }

    public static class EditPhoneNumberFragment extends AbsSignUpPageFragment {
        private EditText mEditPhoneNumber;
        private EditText mEditCountryCode;

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.fragment_sign_up_phone_number, container, false);
        }

        @Override
        protected void updateNextButton() {
            if (!getUserVisibleHint()) return;
            final SignUpActivity signUpActivity = getSignUpActivity();
            if (signUpActivity == null || mEditPhoneNumber == null || mEditCountryCode == null)
                return;
            signUpActivity.setNextEnabled(mEditPhoneNumber.length() > 0 && mEditCountryCode.length() > 0);
        }

        @Override
        public void onViewCreated(final View view, @Nullable final Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            mEditPhoneNumber = (EditText) view.findViewById(R.id.edit_phone_number);
            mEditCountryCode = (EditText) view.findViewById(R.id.edit_country_code);
        }

        @Override
        public void onNextPage() {
            final String phoneNumber = ParseUtils.parseString(mEditPhoneNumber.getText());
            final String countryCode = ParseUtils.parseString(mEditCountryCode.getText());
            getSignUpActivity().sendVerifyCode(phoneNumber, countryCode);
        }

        @Override
        public void onActivityCreated(@Nullable final Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);
            mEditPhoneNumber.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    updateNextButton();
                }

                @Override
                public void afterTextChanged(Editable s) {

                }
            });
            mEditCountryCode.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {

                }

                @Override
                public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
                    updateNextButton();
                }

                @Override
                public void afterTextChanged(final Editable s) {

                }
            });
        }
    }

    public static class VerifyPhoneNumberFragment extends AbsSignUpPageFragment {

        private EditText mEditVerifyCode;

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.fragment_sign_up_verify_phone, container, false);
        }

        @Override
        public void onViewCreated(final View view, @Nullable final Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            mEditVerifyCode = (EditText) view.findViewById(R.id.edit_verify_code);
        }

        @Override
        public void onActivityCreated(@Nullable final Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);
            mEditVerifyCode.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {

                }

                @Override
                public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
                    updateNextButton();
                }

                @Override
                public void afterTextChanged(final Editable s) {

                }
            });
        }

        @Override
        protected void updateNextButton() {
            if (!getUserVisibleHint()) return;
            final SignUpActivity signUpActivity = getSignUpActivity();
            if (signUpActivity == null || mEditVerifyCode == null)
                return;
            signUpActivity.setNextEnabled(mEditVerifyCode.length() > 0);
        }

        @Override
        public void onNextPage() {
            final String verifyCode = ParseUtils.parseString(mEditVerifyCode.getText());
            getSignUpActivity().verifyPhoneNumber(verifyCode);

        }
    }

    public static class EditAvatarFragment extends AbsSignUpPageFragment {


        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.fragment_sign_up_edit_avatar, container, false);
        }

        @Override
        protected void updateNextButton() {
            if (!getUserVisibleHint()) return;
            final SignUpActivity signUpActivity = getSignUpActivity();
            if (signUpActivity == null) return;
            signUpActivity.setNextEnabled(true);
        }

        @Override
        public void onNextPage() {
            final SignUpActivity signUpActivity = getSignUpActivity();
            final Intent data = new Intent();
            signUpActivity.finishAddAddAccount();
        }

    }

    private void finishAddAddAccount() {
        if (mAccessToken == null) return;
        final Intent data = new Intent();
        try {
            data.putExtra(EXTRA_TOKEN, LoganSquare.serialize(mAccessToken));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        setResult(RESULT_OK, data);
        finish();
    }
}

File: src/main/kotlin/catchla/yep/activity/SignUpActivity.kt
/*
 * Copyright 2012 Roman Nurik
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package catchla.yep.activity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.support.v4.view.ViewPager
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.Toast
import catchla.yep.Constants
import catchla.yep.R
import catchla.yep.adapter.TabsAdapter
import catchla.yep.fragment.ProgressDialogFragment
import catchla.yep.model.AccessToken
import catchla.yep.model.Client
import catchla.yep.model.CreateRegistrationResult
import catchla.yep.model.YepException
import catchla.yep.util.YepAPIFactory
import kotlinx.android.synthetic.main.activity_sign_in_sign_up.*
import me.philio.pinentry.PinEntryView
import nl.komponents.kovenant.task
import nl.komponents.kovenant.ui.alwaysUi
import nl.komponents.kovenant.ui.failUi
import nl.komponents.kovenant.ui.successUi

class SignUpActivity : ContentActivity(), Constants, ViewPager.OnPageChangeListener, View.OnClickListener {


    private var mName: String? = null
    private var mCreateRegistrationResult: CreateRegistrationResult? = null
    private var mAccessToken: AccessToken? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_in_sign_up)
        val adapter = TabsAdapter(this, supportFragmentManager)
        viewPager.adapter = adapter
        viewPager.isEnabled = false
        viewPager.addOnPageChangeListener(this)

        nextButton.setOnClickListener(this)

        adapter.addTab(EditNameFragment::class.java, null, 0, null)
        adapter.addTab(EditPhoneNumberFragment::class.java, null, 0, null)
        adapter.addTab(VerifyPhoneNumberFragment::class.java, null, 0, null)
        adapter.addTab(EditAvatarFragment::class.java, null, 0, null)
    }

    override fun onContentChanged() {
        super.onContentChanged()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

    }

    override fun onPageSelected(position: Int) {

    }

    override fun onPageScrollStateChanged(state: Int) {

    }

    override fun onClick(v: View) {
        when (v) {
            nextButton -> {
                val fragment = (viewPager.adapter as TabsAdapter).primaryItem
                if (fragment is AbsSignUpPageFragment) {
                    fragment.onNextPage()
                }
            }
        }
    }

    private fun setName(name: String) {
        mName = name
    }

    private fun setNextEnabled(enabled: Boolean) {
        nextButton.isEnabled = enabled
    }

    private fun gotoNextPage() {
        val currentItem = viewPager.currentItem
        if (currentItem == viewPager.adapter.count - 1) return
        viewPager.currentItem = currentItem + 1
    }

    private fun sendVerifyCode(phoneNumber: String, countryCode: String) {
        ProgressDialogFragment.show(this, "create_registration")
        task {
            val yep = YepAPIFactory.getInstanceWithToken(this@SignUpActivity, null)
            yep.createRegistration(phoneNumber, countryCode, mName!!, 0.0, 0.0)
        }.successUi {
            setCreateRegistrationResult(it)
            gotoNextPage()
        }.failUi {
            @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
            val exception = it as? YepException
            val error = exception?.error
            if (error != null && !TextUtils.isEmpty(error)) {
                Toast.makeText(this, error, Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, R.string.unable_to_register, Toast.LENGTH_SHORT).show()
            }
            val fragment = currentFragment
            if (fragment is VerifyPhoneNumberFragment) {
                fragment.clearPin()
            }
        }.alwaysUi {
            val f = supportFragmentManager.findFragmentByTag("create_registration")
            if (f is DialogFragment) {
                f.dismiss()
            }
        }

    }

    private val currentFragment: Fragment?
        get() = viewPager.adapter.instantiateItem(viewPager, viewPager.currentItem) as Fragment?

    private fun verifyPhoneNumber(verifyCode: String) {
        ProgressDialogFragment.show(this, "update_registration")
        task {
            val yep = YepAPIFactory.getInstanceWithToken(this@SignUpActivity, null)
            val result = mCreateRegistrationResult!!
            yep.updateRegistration(result.mobile, result.phoneCode, verifyCode, Client.OFFICIAL, 0)
        }.successUi {
            setUpdateRegistrationResult(it)
            gotoNextPage()
        }.failUi {
            @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
            val exception = it as? YepException
            val error = exception?.error
            if (TextUtils.isEmpty(error)) {
                Toast.makeText(this, R.string.unable_to_verify_phone, Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, error, Toast.LENGTH_SHORT).show()
            }
        }.alwaysUi {
            val f = supportFragmentManager.findFragmentByTag("update_registration")
            if (f is DialogFragment) {
                f.dismiss()
            }
        }
    }

    private fun setUpdateRegistrationResult(result: AccessToken) {
        mAccessToken = result
    }

    fun setCreateRegistrationResult(result: CreateRegistrationResult) {
        mCreateRegistrationResult = result
    }

    abstract class AbsSignUpPageFragment : Fragment() {
        protected abstract fun updateNextButton()

        abstract fun onNextPage()

        override fun setUserVisibleHint(isVisibleToUser: Boolean) {
            super.setUserVisibleHint(isVisibleToUser)
            if (isVisibleToUser) {
                updateNextButton()
            }
        }

        val signUpActivity: SignUpActivity?
            get() = activity as SignUpActivity

        override fun onActivityCreated(savedInstanceState: Bundle?) {
            super.onActivityCreated(savedInstanceState)
            updateNextButton()
        }
    }

    class EditNameFragment : AbsSignUpPageFragment() {
        private var mEditName: EditText? = null

        override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater!!.inflate(R.layout.fragment_sign_up_edit_name, container, false)
        }

        override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            mEditName = view!!.findViewById(R.id.edit_name) as EditText
        }

        override fun onActivityCreated(savedInstanceState: Bundle?) {
            super.onActivityCreated(savedInstanceState)
            mEditName!!.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    updateNextButton()
                }

                override fun afterTextChanged(s: Editable) {

                }
            })
        }

        override fun updateNextButton() {
            if (!userVisibleHint) return
            val signUpActivity = signUpActivity
            if (signUpActivity == null || mEditName == null) return
            signUpActivity.setNextEnabled(mEditName!!.length() > 0)
        }

        override fun onNextPage() {
            val signUpActivity = signUpActivity
            signUpActivity!!.setName(mEditName!!.text.toString())
            signUpActivity.gotoNextPage()
        }
    }

    class EditPhoneNumberFragment : AbsSignUpPageFragment() {
        private var mEditPhoneNumber: EditText? = null
        private var mEditCountryCode: EditText? = null

        override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater!!.inflate(R.layout.fragment_sign_up_phone_number, container, false)
        }

        override fun updateNextButton() {
            if (!userVisibleHint) return
            val signUpActivity = signUpActivity
            if (signUpActivity == null || mEditPhoneNumber == null || mEditCountryCode == null)
                return
            signUpActivity.setNextEnabled(mEditPhoneNumber!!.length() > 0 && mEditCountryCode!!.length() > 0)
        }

        override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            mEditPhoneNumber = view!!.findViewById(R.id.edit_phone_number) as EditText
            mEditCountryCode = view.findViewById(R.id.edit_country_code) as EditText
        }

        override fun onNextPage() {
            val phoneNumber = mEditPhoneNumber!!.text.toString()
            val countryCode = mEditCountryCode!!.text.toString()
            signUpActivity!!.sendVerifyCode(phoneNumber, countryCode)
        }

        override fun onActivityCreated(savedInstanceState: Bundle?) {
            super.onActivityCreated(savedInstanceState)
            mEditPhoneNumber!!.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    updateNextButton()
                }

                override fun afterTextChanged(s: Editable) {

                }
            })
            mEditCountryCode!!.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    updateNextButton()
                }

                override fun afterTextChanged(s: Editable) {

                }
            })
        }
    }

    class VerifyPhoneNumberFragment : AbsSignUpPageFragment() {

        private var mEditVerifyCode: PinEntryView? = null

        override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater!!.inflate(R.layout.fragment_sign_up_verify_phone, container, false)
        }

        override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            mEditVerifyCode = view!!.findViewById(R.id.edit_verify_code) as PinEntryView
        }

        override fun onActivityCreated(savedInstanceState: Bundle?) {
            super.onActivityCreated(savedInstanceState)
            mEditVerifyCode!!.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    updateNextButton()
                    if (verifyPin()) {
                        onNextPage()
                    }
                }

                override fun afterTextChanged(s: Editable) {

                }
            })
        }

        override fun updateNextButton() {
            if (!userVisibleHint) return
            val signUpActivity = signUpActivity
            if (signUpActivity == null || mEditVerifyCode == null)
                return
            signUpActivity.setNextEnabled(verifyPin())
        }

        override fun onNextPage() {
            val verifyCode = mEditVerifyCode!!.text.toString()
            signUpActivity!!.verifyPhoneNumber(verifyCode)

        }

        private fun verifyPin(): Boolean {
            return mEditVerifyCode!!.text.length == 4
        }


        fun clearPin() {
            mEditVerifyCode!!.clearText()
        }
    }

    class EditAvatarFragment : AbsSignUpPageFragment() {


        override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater!!.inflate(R.layout.fragment_sign_up_edit_avatar, container, false)
        }

        override fun updateNextButton() {
            if (!userVisibleHint) return
            val signUpActivity = signUpActivity ?: return
            signUpActivity.setNextEnabled(true)
        }

        override fun onNextPage() {
            val signUpActivity = signUpActivity
            val data = Intent()
            signUpActivity!!.finishAddAddAccount()
        }

    }

    private fun finishAddAddAccount() {
        if (mAccessToken == null) return
        val data = Intent()
        data.putExtra(Constants.EXTRA_TOKEN, mAccessToken)
        setResult(Activity.RESULT_OK, data)
        finish()
    }
}